### PR TITLE
Addon Test: Refactor test addon to include stories automatically

### LINKED ISSUE #29771: [Bug] some stories are visible in the index but not in Vitest, or vice versa
### [stories glob](https://storybook.js.org/docs/api/main-config/main-config-stories)

Currently, we rely on the `include` property of the Vitest config file e.g.

```tsx
// vitest.workspace.js
export default defineWorkspace([
	{ 
	  plugins: [ 
	    storybookTest(),
	  ],
	  include: ['**/*.stories.?(m)[jt]s?(x)'],
	  // ...
	}
])
```

Because this is using an arbitrary include glob and does not related to the one in main.js, stories can be tested even though they are ignored in Storybook, or vice versa.

### PR DESCRIPTION
**Pull request by yannbf**
Closes #29771

## What I did

This PR introduces a change in the Vitest plugin where it leverages the `stories` field of `main.ts` to transform into an array of files for `vitest`, and adds that automatically as `config.test.include`.

I tested this locally in the monorepo's Storybook and it worked without issues. Adding extra `include` or `exclude` paths in the `vitest` config will work as expected.

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [x] integration tests
- [ ] end-to-end tests

#### Manual testing

This is tested by our sandboxes, as well as via our `storybook:vitest` npm script; Which is run in CI.

You can experiment by adding a new story to the monorepo storybook, and running `storybook:vitest` and check if the new story is tested.

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createSize |  0 B | 0 B | 0 B | - | - |
| generateSize |  77.7 MB | 77.7 MB | 0 B | 0.68 | 0% |
| initSize |  130 MB | 130 MB | 4.88 kB | 0.3 | 0% |
| diffSize |  52.4 MB | 52.4 MB | 4.88 kB | 0.21 | 0% |
| buildSize |  6.75 MB | 6.75 MB | 108 B | **-1.36** | 0% |
| buildSbAddonsSize |  1.51 MB | 1.51 MB | 0 B | - | 0% |
| buildSbCommonSize |  195 kB | 195 kB | 0 B | - | 0% |
| buildSbManagerSize |  1.86 MB | 1.86 MB | 108 B | 1 | 0% |
| buildSbPreviewSize |  0 B | 0 B | 0 B | **-1.36** | - |
| buildStaticSize |  0 B | 0 B | 0 B | - | - |
| buildPrebuildSize |  3.56 MB | 3.57 MB | 108 B | **-1.36** | 0% |
| buildPreviewSize |  3.19 MB | 3.19 MB | 0 B | **1.36** | 0% |
| testBuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbAddonsSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbCommonSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbManagerSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| testBuildStaticSize |  0 B | 0 B | 0 B | - | - |
| testBuildPrebuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildPreviewSize |  0 B | 0 B | 0 B | - | - |

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createTime |  23.9s | 6.8s | -17s -94ms | -1.06 | -249.9% |
| generateTime |  19.2s | 21.1s | 1.9s | 0.34 | 9.1% |
| initTime |  12.8s | 14.2s | 1.3s | 0.29 | 9.6% |
| buildTime |  9.9s | 10.5s | 541ms | 1.21 | 5.1% |
| testBuildTime |  0ms | 0ms | 0ms | - | - |
| devPreviewResponsive |  5.7s | 4.2s | -1s -467ms | **-1.56** | 🔰-34.7% |
| devManagerResponsive |  4.2s | 3.2s | -1s -13ms | **-1.26** | 🔰-31.5% |
| devManagerHeaderVisible |  655ms | 504ms | -151ms | -1.06 | -30% |
| devManagerIndexVisible |  726ms | 568ms | -158ms | -1.02 | -27.8% |
| devStoryVisibleUncached |  2s | 1.7s | -351ms | -0.28 | -20.4% |
| devStoryVisible |  727ms | 532ms | -195ms | **-1.25** | 🔰-36.7% |
| devAutodocsVisible |  605ms | 489ms | -116ms | -0.69 | -23.7% |
| devMDXVisible |  643ms | 444ms | -199ms | -0.93 | -44.8% |
| buildManagerHeaderVisible |  674ms | 494ms | -180ms | -0.9 | -36.4% |
| buildManagerIndexVisible |  792ms | 569ms | -223ms | -0.61 | -39.2% |
| buildStoryVisible |  617ms | 460ms | -157ms | -1.09 | -34.1% |
| buildAutodocsVisible |  566ms | 422ms | -144ms | -0.68 | -34.1% |
| buildMDXVisible |  520ms | 380ms | -140ms | -1.16 | -36.8% |

## Greptile Summary

Based on the provided files and context, here's a concise summary of the changes:

Refactored the Storybook test addon to automatically sync story discovery between Storybook and Vitest testing.

- Removed hardcoded story glob patterns from Vitest config in favor of using Storybook's `stories` field from main.js
- Added `StoryIndexGenerator` export from core-server for consistent story file discovery
- Added automatic MDX file exclusion and framework-specific plugin configurations for Next.js, SvelteKit, and Vue3
- Fixed setup file path resolution in workspace configuration for better monorepo support
- Added framework-specific plugin imports and configurations to handle different framework requirements

The changes aim to resolve inconsistencies where stories might be visible in Storybook but not in Vitest tests or vice versa, while maintaining compatibility with different frameworks and build configurations.

### CODE REVIEW

File: code/.storybook/vitest.config.ts [hunk 1]
```diff
@@ -34,20 +34,15 @@ export default mergeConfig(
     ],
     test: {
       name: 'storybook-ui',
-      include: [
-        '../addons/**/*.{story,stories}.?(c|m)[jt]s?(x)',
-        // '../core/template/stories/**/*.{story,stories}.?(c|m)[jt]s?(x)',
-        '../core/src/manager/**/*.{story,stories}.?(c|m)[jt]s?(x)',
-        '../core/src/preview-api/**/*.{story,stories}.?(c|m)[jt]s?(x)',
-        '../core/src/components/{brand,components}/**/*.{story,stories}.?(c|m)[jt]s?(x)',
-      ],
```
Comment (ndelangen) [inline code/.storybook/vitest.config.ts:43]: We should no longer do this... as now we determine which stories are to be loaded as tests, via the index.

File: code/.storybook/vitest.config.ts [hunk 2]
```diff
@@ -34,20 +34,15 @@ export default mergeConfig(
...
       exclude: [
         ...defaultExclude,
         '../node_modules/**',
         '**/__mockdata__/**',
         '../**/__mockdata__/**',
         // expected to fail in Vitest because of fetching /iframe.html to cause ECONNREFUSED
         '**/Zoom.stories.tsx',
+        // @yannbf what's wrong with these stories, that they do not work
+        '**/lib/blocks/src/**',
```
[the remark this answers is not in the document]
Comment (yannbf) [reply code/.storybook/vitest.config.ts:45]: They won't work because of https://github.com/storybookjs/storybook/issues/29783

**Review by ndelangen — APPROVED** [review]

**Review by ghengeveld — APPROVED** [review]

**Comment by JReinhold** [comment]
@ghengeveld @ndelangen  I wonder if our new way of finding stories here, also fixes https://github.com/storybookjs/storybook/issues/29572. We should try it out in QA.

**Comment by yannbf** [comment]
> @ghengeveld @ndelangen I wonder if our new way of finding stories here, also fixes #29572. We should try it out in QA.

It doesn't seem to fix it. Here's an example on windows with two directories: `my stories` and `my tests` with a simple Vitest file. Vitest is able to find and test the `my tests` one, but fails in `my stories`:
image

strangely enough if I use the newest alpha and I keep the include which was there originally (the default one with *.stories.*), then the tests fail because they end up finding the template stories in node_modules
image

**Pull request merged**